src/implementations/junoswap/pool.rs

```diff
@@ -0,0 +1,337 @@
+            pool_info.token1_reserve,
+        )?;
+
+        // TODO: Is this the behavior of slippage_tolerance that we want? Right now
+        // It's a bit unclear what slippage_tolerance is supposed to do. We must
+        // define it more clearly in the trait doc comments.
+        let min_liquidity = expected_lps
+            * Decimal::one().checked_sub(slippage_tolerance.unwrap_or_else(|| Decimal::one()))?;
+
+        // Increase allowance for cw20 tokens and add native tokens to the funds vec.
+        let (funds, increase_allowances) =
+            prepare_funds_and_increase_allowances(env, info, assets, &self.addr)?;
+
+        let provide_liquidity = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: self.addr.to_string(),
+            funds,
+            msg: to_binary(&ExecuteMsg::AddLiquidity {
+                token1_amount: token1_to_use,
+                min_liquidity,
+                max_token2: token2_to_use,
+                expiration: None,
+            })?,
+        });
+
+        let event = Event::new("apollo/cw-dex/provide_liquidity")
+            .add_attribute("type", "junoswap")
+            .add_attribute("assets", format!("{:?}", vec![token1, token2]))
+            .add_attribute("recipient", recipient.to_string());
+
+        Ok(Response::new()
+            .add_messages(increase_allowances)
+            .add_message(provide_liquidity)
+            .add_event(event))
+    }
+
+    fn withdraw_liquidity(
+        &self,
+        _deps: Deps,
+        asset: Asset,
+        recipient: Addr,
+    ) -> Result<Response, CwDexError> {
+        let withdraw_liquidity = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: self.addr.to_string(),
+            funds: vec![],
+            msg: to_binary(&ExecuteMsg::RemoveLiquidity {
+                amount: asset.amount,
+                min_token1: Uint128::zero(),
+                min_token2: Uint128::zero(),
+                expiration: None,
+            })?,
+        });
+
+        let event = Event::new("apollo/cw-dex/withdraw_liquidity")
+            .add_attribute("type", "junoswap")
+            .add_attribute("asset", format!("{:?}", asset))
+            .add_attribute("recipient", recipient.to_string());
+
+        Ok(Response::new().add_message(withdraw_liquidity).add_event(event))
+    }
+
+    fn swap(
+        &self,
+        deps: Deps,
+        offer_asset: Asset,
+        ask_asset_info: AssetInfo,
+        minimum_out_amount: Uint128,
+        recipient: Addr,
+    ) -> Result<Response, CwDexError> {
+        let pool_info = self.query_info(&deps.querier)?;
+
+        let input_token = if JunoAssetInfo(pool_info.token1_denom) == offer_asset.info {
+            Ok(TokenSelect::Token1)
+        } else if JunoAssetInfo(pool_info.token2_denom) == offer_asset.info {
+            Ok(TokenSelect::Token2)
+        } else {
+            Err(StdError::generic_err("Offered asset is not in the pool"))
+        }?;
```

Comment:
Should we check that the ask_asset_info is the other token in the pool for safety? Like is done in `simulate_swap`?